### ISSUE TITLE
Apathy seems to rely on setsockopt(), which may be unreliable on Linux?

### ISSUE THREAD

**Issue by crowder-pw** [vis2k/DOTSNET#1]
**Describe the bug**

On Linux, setsockopt(socket, SOL_SOCKET, SO_SNDBUF, 7340032, 4) can succeed, but NOT yield a send buffer that is actually 7 * 1024 * 1024 bytes in size. To demonstrate this, I have added the following code to Common.cs, immediately after the network_set_send_buffer_size() call:

```
        int realsize = 0;
        if (NativeBindings.network_get_send_buffer_size(socket, ref realsize, ref error) != 0)
            Debug.LogError("network_get_send_buffer_size faioled: " + (NativeError)error);

        if (realsize != SendBufferSize)
            Debug.LogWarning($"OS send buffer is not the size requested: {realsize} != {SendBufferSize}");
```

In strace(), the resulting output looks like this:

```
0.000178 setsockopt(6, SOL_SOCKET, SO_SNDBUF, [7340032], 4) = 0
0.000240 getsockopt(6, SOL_SOCKET, SO_SNDBUF, [425984], [4]) = 0
```

Notice that the getsockopt() result value is 425984.

This results in Apathy overflowing the socket's send buffer much more quickly than it expects:

```
0.001279 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\376*\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000500 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\335&\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000370 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\274\"\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000341 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\233\36\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000350 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254z\32\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000349 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254Y\26\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000334 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\2548\22\0\0\1\0"..., 65538, 0, NULL, 0) = 65538
0.000335 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\27\16\0\0\1\0"..., 65538, 0, NULL, 0) = 5514
0.000360 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\366\t\0\0\1\0"..., 65538, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
```

**To Reproduce**

Run a DOTSNET project on Linux with the Common.cs patch shown above, observe the following output:

`
OS send buffer is not the size requested: 425984 != 7340032
`

**Expected behavior**

This is probably a feature request, since I "expect" that Apathy not rely on the OS send buffer and instead do its own buffering. Or, I need to figure out how to throttle/buffer at a higher level.

**Desktop (please complete the following information):**
 - OS: Ubuntu Linux (5.3.0-1019-aws)

**Comment by crowder-pw** [comment]
Another possible bug worth calling out here:

One of the strace lines from above...
```
0.000335 sendto(11, "\0\0\377\376:\0\0\0\"\0\246\302$\370\2518\304?\252\332>\330\233\23\306\254\27\16\0\0\1\0"..., 65538, 0, NULL, 0) = 5514
```

This suggests that there was enough buffer-space remaining here to send 5514 bytes. The sentto() did not FAIL, but did send less than requested. If we do not want clients to disconnect here (maybe we do??) then we need to keep the remaining bytes available for a future send. If we DO want clients to disconnect as a result of these buffers filling, then we probably need to consider an amount-sent that's less than the amount intended as an error, without waiting for the EAGAIN (which the Apathy and DOTSNET code treat as EWOULDBLOCK).

**Comment by vis2k** [comment]
Thanks for the report.

Added a 'set large send/recv buffer test' to Apathy so the problem is at least more obvious to detect/debug for now.

ConfigureSocket() now double checks the internal size after setting the buffers and logs an error if failed. This way we see the problem immediately instead of having strange suprises at rumtime.

Next step is figuring out how to deal with the situation now that it is detected properly.

**Comment by vis2k** [comment]
Next change: Apathy send/recv buffer sizes are now reduced from 7MB to 416KB on all platforms because that's the minimum value that works on Windows/Mac/Linux.

They were also made const so that no one accidentally increases them, as the linux max size is not obvious at all.

**Comment by vis2k** [comment]
Next change: added a full 416KB send test while verifying every single sent/received byte to make sure it works.
Turns out we can only successfully send 363KB on linux.

**Comment by vis2k** [comment]
Fixed in Apathy V1.12.
Thanks for reporting guys!
